### DISCUSSION TITLE
What is Deno-based scripting system?

### DISCUSSION THREAD

**Discussion by madwareru** [fishfolk/jumpy#468] [Q&A] [answered]
Hey! I heard from @erlend-sh in bevy discord that there is plans to use some kind of scripting system based on Deno. Am I right if I think you are planning to use nodejs alternative named Deno, or is it just coincidence? Where I can find additional info on it?

**Comment by erlend-sh** [comment]
Jumpy will indeed also use this system, but it’s gonna land first in Punchy: https://github.com/fishfolks/punchy/pull/255

It’s based on this: https://github.com/jakobhellermann/bevy_mod_js_scripting

**Comment by zicklag** [comment, marked as answer]
Hi @madwareru! We are using [Deno](https://deno.land)'s core wrapper around V8, which is the underlying C++ JavaScript engine that Node.js, Deno, and Chrome use.

It doesn't have any of Deno's runtime features, though, such as filesystem or network access. The scripts can only access the ECS game world, which is actually great for running untrusted mods.

Like @erlend-sh said, we are using the [`bevy_mod_js_scripting`](https://github.com/jakobhellermann/bevy_mod_js_scripting) plugin to integrate Deno with Bevy.

It's also worth noting that when running on the web Deno isn't used, but the native JavaScript engine of the browser playing the game is used instead.

Right now `bevy_mod_js_scripting` is very work-in-progress, but it's developing nicely. As I'm working on integrating scripting with Punchy and Jumpy, I'm adding the extra features we need to `bevy_mod_js_scripting` while working with @jakobhellermann to figure out the best way to design the scripting API.

There aren't a lot of other docs yet, but you can see the [Punchy Wiki page on Scripting](https://github.com/fishfolks/punchy/wiki/Scripting) to get some of an idea of how we want to use it. You can also check out the [`demo_script.ts`](https://github.com/fishfolks/punchy/blob/000ae6d58f289cb161624fe6a521718e5330dd36/assets/scripts/demo_script.ts) and [`health.ts`](https://github.com/fishfolks/punchy/blob/000ae6d58f289cb161624fe6a521718e5330dd36/assets/items/health/health.ts) scripts to get an idea for what the API is like.

Also, I'd be happy to answer any questions you have.
